R/srcrefs.R

```diff
@@ -139,17 +139,47 @@ srcrefs.R6ClassGenerator <- function(x, ..., srcref_names = NULL,
 
 ### S4
 
+#' @exportS3Method
+#' @importFrom utils getSrcref
+#' @rdname srcrefs
+srcrefs.standardGeneric <- function(x, ..., srcref_names = NULL) {
+  if (is.null(sr <- getSrcref(x))) return(sr)
+
+  attr(sr, "namespace") <- obj_namespace_name(x)
+
+  objs <- list(sr)
+  names(objs) <- srcref_names
+  objs
+}
+
+#' @exportS3Method
+#' @importFrom utils getSrcref
+#' @rdname srcrefs
+srcrefs.nonstandardGenericFunction <- srcrefs.standardGeneric
+
 #' @exportS3Method
 #' @importFrom utils getSrcref
 #' @rdname srcrefs
 srcrefs.MethodDefinition <- function(x, ..., srcref_names = NULL) {
   # catch methods in methods tables from packages without srcref data
   if (is.null(sr <- getSrcref(x))) return(sr)
 
-  # as produced by `methods:::.methods_info`
-  signatures <- paste0(x@generic, ",", paste0(x@defined, collapse = ","), "-method")
+  # generic source package
+  generic_origin_ns <- attr(x@generic, "package")
+  ns <- env_ns_name(environment(x@.Data))
+
+  # if method is defined in same package as generic, use generic name as
+  # signature alias, otherwise use methods_info-style method alias name
+  if (generic_origin_ns == ns) {
+    signatures <- x@generic
+  } else {
+    # as produced by `methods:::.methods_info`
+    signatures <- paste0(x@generic, ",", paste0(x@defined, collapse = ","), "-method")
+  }
```

Comment:
Discovered along the way, generics that live outside the target package follow a different alias naming convention than those from within.

Reply:
Wow, that would be a real nightmare to debug had it become an issue at one point! Great job